app/test/consistent_apphash_test.go

```diff
@@ -3,134 +3,303 @@ package app_test
-		accounts = append(accounts, account)
-	}
+	accounts := createAccounts(accountInfos, accountNames)
 
-	// Create a signer with keyring accounts
+	// Create a signer with accounts
 	signer, err := user.NewSigner(kr, enc.TxConfig, testutil.ChainID, app.DefaultInitialVersion, accounts...)
 	require.NoError(t, err)
 
-	amount := sdk.NewCoins(sdk.NewCoin(app.BondDenom, sdk.NewIntFromUint64(1000)))
+	// ------------ Genesis Validator Accounts  ------------
 
-	// Create an SDK Tx
-	sdkTx := SdkTx{
-		sdkMsgs: []sdk.Msg{
-			banktypes.NewMsgSend(signer.Account(accountNames[0]).Address(),
-				signer.Account(accountNames[1]).Address(),
-				amount),
-		},
-		txOptions: blobfactory.DefaultTxOpts(),
+	// Validators from genesis state
+	genValidators := testApp.StakingKeeper.GetAllValidators(testApp.NewContext(false, tmproto.Header{}))
+
+	// Get validator account names from the validator keyring
+	valAccountNames := testfactory.GetAccountNames(valKeyRing)
+
+	// Query validator account infos
+	valAccountInfos := queryAccountInfo(testApp, valAccountNames, valKeyRing)
+
+	// Create accounts for the validators' signer
+	valAccounts := createAccounts(valAccountInfos, valAccountNames)
+
+	// Create a signer with validator accounts
+	valSigner, err := user.NewSigner(valKeyRing, enc.TxConfig, testutil.ChainID, app.DefaultInitialVersion, valAccounts...)
+	require.NoError(t, err)
+
+	// ----------- Create SDK Messages ------------
+
+	amount := sdk.NewCoins(sdk.NewCoin(app.BondDenom, sdk.NewIntFromUint64(1_000)))
+	// Minimum deposit required for a gov proposal to become active
+	depositAmount := sdk.NewCoins(sdk.NewCoin(app.BondDenom, sdk.NewIntFromUint64(10000000000)))
+	twoInt := sdk.NewInt(2)
+
+	// ---------------- First Block ------------
+	var firstBlockSdkMsgs []sdk.Msg
+
+	// NewMsgSend - sends funds from account-0 to account-1
+	sendFundsMsg := banktypes.NewMsgSend(accountAddresses[0], accountAddresses[1], amount)
+	firstBlockSdkMsgs = append(firstBlockSdkMsgs, sendFundsMsg)
+
+	// MultiSend - creates a multi-send transaction from account-0 to account-1
+	multiSendFundsMsg := banktypes.NewMsgMultiSend([]banktypes.Input{
+		banktypes.NewInput(
+			accountAddresses[0],
+			amount,
+		),
+	},
+		[]banktypes.Output{
+			banktypes.NewOutput(
+				accountAddresses[1],
+				amount,
+			),
+		})
+	firstBlockSdkMsgs = append(firstBlockSdkMsgs, multiSendFundsMsg)
+
+	// NewMsgGrant - grants authorization to account-1
+	grantExpiration := time.Date(2026, time.January, 1, 0, 0, 0, 0, time.UTC)
+	authorization := authz.NewGenericAuthorization(blobtypes.URLMsgPayForBlobs)
+	msgGrant, err := authz.NewMsgGrant(
+		accountAddresses[0],
+		accountAddresses[1],
+		authorization,
+		&grantExpiration,
+	)
+	require.NoError(t, err)
+	firstBlockSdkMsgs = append(firstBlockSdkMsgs, msgGrant)
+
+	// MsgVerifyInvariant - verifies the nonnegative-outstanding invariant within the bank module for the account-0
```

Comment:
[tangent] should we have specs for all supported msg types? Could be a follow-up issue.

Reply:
https://github.com/celestiaorg/celestia-app/issues/3678